src/Routing/PushState.purs

```diff
@@ -0,0 +1,247 @@
+  , listen :: (LocationState -> Eff eff Unit) -> Eff eff (Eff eff Unit)
+  }
+
+-- | Creates a new `PushStateInterface` given a `Basename`. The `Basename` is
+-- | used as an implicit prefix for the paths supplied to the instance, being
+-- | automatically added and stripped as necessary.
+makeInterface :: forall eff. Basename -> Eff (PushStateEffects eff) (PushStateInterface (PushStateEffects eff))
+makeInterface basename = do
+  argsRef <- newRef []
+  freshRef <- newRef 0
+  listenersRef <- newRef StrMap.empty
+
+  let
+    notify ev =
+      traverse_ (_ $ ev) =<< readRef listenersRef
+
+    listen k = do
+      fresh <- readRef freshRef
+      writeRef freshRef (fresh + 1)
+      modifyRef listenersRef $ StrMap.insert (show fresh) k
+      pure $ modifyRef listenersRef $ StrMap.delete (show fresh)
+
+    locationState = do
+      loc <- DOM.window >>= DOM.location
+      state <- DOM.window >>= DOM.history >>= DOM.state
+      pathname <- fromMaybe <*> String.stripPrefix (String.Pattern (unwrap basename)) <$> DOM.pathname loc
+      search <- DOM.search loc
+      hash <- DOM.hash loc
+      let path = pathname <> search <> hash
+      pure { state, pathname, search, hash, path }
+
+  -- The hashchange interface is asynchronous, since hashchange events are
+  -- fired on the next tick of the event loop. We want the push-state
+  -- interface to behave as similarly as possible, so we use the microtask
+  -- queue via MutationObserver to schedule callbacks. Alternatively we could
+  -- just use a setTimeout, but it would not be as prompt. We use a fresh
+  -- counter so that the text change mutation always fires.
+  schedule <- do
+    obsvNode <-
+      DOM.window
+        >>= DOM.document
+        >>> map DOM.htmlDocumentToDocument
+        >>= DOM.createTextNode ""
+        >>> map DOM.textToNode
+    observer <- DOM.mutationObserver \_ _ -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef []
+      for_ argsQueue notify
+    DOM.observe obsvNode { characterData: true } observer
+    pure \args -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef (Array.snoc argsQueue args)
+      when (Array.null argsQueue) do
+        fresh ← readRef freshRef
+        writeRef freshRef (fresh + 1)
+        DOM.setNodeValue (show fresh) obsvNode
+
+  let
+    stateFn op state path = do
+      let
+        loc = normalizeLocation state path
+        url = DOM.URL $ unwrap basename <> loc.pathname <> loc.search <> loc.hash
+      DOM.window
+        >>= DOM.history
+        >>= op state (DOM.DocumentTitle "") url
+      schedule loc
+
+    listener =
+      DOM.eventListener \_ -> notify =<< locationState
+
+  DOM.window
+    >>= DOM.windowToEventTarget
+    >>> DOM.addEventListener DOM.popstate listener false
+
+  pure
+    { pushState: stateFn DOM.pushState
+    , replaceState: stateFn DOM.replaceState
+    , locationState
+    , listen
+    }
+
+type LocationState =
+  { state :: Foreign
+  , path :: String
```

Comment:
:+1:

Reply:
I did not because there is already a key `path` to distinguish it.

Reply:
I'm trying to reflect the properties returned by the DOM API, which are all strings. Maybe I should remove the `Path` newtype? If I'm removing `Basename` then I don't see why I should keep it at all since there's no ambiguity with anything.

Reply:
:+1: for strings if there is no ambiguity